### ISSUE TITLE
Kotlin/Native support

### ISSUE THREAD

**Issue by eyelash** [JetBrains/Exposed#635]
Are there any plans to support Kotlin/Native?

**Comment by Tapac** [comment]
It's not on the roadmap right now but we could consider it after implementing support for async java drivers.

**Comment by vkhooda24** [comment]
@Tapac I really love this library and using with Ktor and Android. I am not sure how much effort it will take to make it compatible with Kotlin multiplatform but certainly, if you consider it to do sooner, it would certainly lead over SQLDelight.
Just imagine, if you are using the same library for server-side and your Android app and get support for multiplatform, how wonderful a dev life would be. :)  Multiplatform already provides libraries for Coroutine, Http client, Serialization support and just missing this one piece for app side database.

**Comment by lbensaad** [comment]
@Tapac It's been a year now, I just wanted to ask if kotlin multiplatform support is on your roadmap now?

**Comment by SeekDaSky** [comment]
Well, a prototype of async support is coming to Exposed just now so my guess is that MPP support isn't coming any time soon.

**Comment by Tapac** [comment]
@lbensaad , sorry, but I don't really have free time to dive into the native world. The problem is that I need to find the native drivers for each database and I'm not sure that there is a common interface such as JDBC in a native world. 

It will be great if someone will try to contribute into the project in place of native support.

**Comment by saied89** [comment]
@Tapac  For MMP sqlite is enough.

**Comment by chris-hatton** [comment]
Just a note to say I came looking; interested in reusing some of my server-side Exposed Repository classes on the client across Android / iOS / Desktop. As @saied89 says, SQLite on iOS is the extend of the K/N support I'd be needing.  Thanks for Exposed, it's a pleasurable API!

**Comment by rustamsmax** [comment]
2 years have passed. I see this feature has appeared on your roadmap. It would be a good alternative to SqlDelight if it supports iOS

**Comment by mcxinyu** [comment]
Is there any progress?

**Comment by mahramane** [comment]
Hi
if Exposed support `kotlin multiplatform (jvm, android, ios, wasm)` then we write one code for server and sqlite cache database in client.
it's will be very easy to develop.
please get help and idea from `sqldelight` library.

**Comment by zoujiaqing** [comment]
Ktor 3.0 will support Kotlin Native.

It's time for Exposed to support Kotlin Native as well.

**Comment by smyrgeorge** [comment]
I faced exactly the same problem, while I was playing around with Ktor 3.0 (beta).
Until now there isn't any mature database driver that target the Kotlin Native.
Although, there quite a few projects that wrap a C driver, using FFI.

I did the same, I created a small driver, wrapping the `sqlx` driver from the Rust ecosystem.
If you want take a look here: https://github.com/smyrgeorge/sqlx4k

**Comment by H4kt** [comment]
Would really love to see exposed for KMP.
I much prefer it to room so I would use Exposed not only on server but in mobile apps too

**Comment by mslalith** [comment]
is there any plan to have on Roadmap soon?

**Comment by smyrgeorge** [comment]
Hello, is there any plan to add support for KMP in the `exposed-core` package?
As I saw the only non-kotlin-native dependency is `slf4j`.
So, I guess will not be that hard (until the reality proves the opposite 😅).

If yes, I could then help with the implementation for the different DB drivers.
As I mentioned in a previous comment I'm currently working on a database driver for Kotlin Native.

**Comment by BierDav** [comment]
Very interessed in this, because I also want to work on an exposed over kotlinx.rpc. And additionally we could use exposed for ios development

**Comment by smyrgeorge** [comment]
Great news!
Whenever you prepare the PR I would also like to take a look.
Then I could start working with the driver integration (in the meantime I will take a look at `exposed-jdbc` module).

The driver that I'm currently implementing wraps the `sqlx` driver from the Rust ecosystem,
thus it only supports PostreSQL, MySQL and SQLite.
I think works quite nice and does not add extra overhead (that was actually my goal).

**Comment by BierDav** [comment]
Sounds good, but I wanted to add that I'm not sure if I am even able to pull this off on myself, because currently I quite stuck on all those weird gradle plugin is applied or not applied or alrady applied, ... things. Quite discouraging, when considering the fact that I don't have really understood how gradle handles this all under hood. But I see it as an opportunity to learn something new, maybe I have luck and get it to work 🫠 keep my fingers crossed

**Comment by smyrgeorge** [comment]
Sure no problem. Also another problem is that Exposed does not support coroutines (at least for the moment).

**Comment by teras** [comment]
I am also looking forward for a way to connect my K/N application to (a) database(s). During my search I found [unixODBC](https://www.unixodbc.org/) project, which, although dated, seems to solve the connectivity problem. I was thinking to give it a go and use this, for my low level DB needs. @smyrgeorge do you believe the `sqlx` driver is a better approach, giving it supports less backends?

At any case, I believe it would be crucial to distinguish the low-level DB drivers from the high level API calls. This would be beneficial if other APIs, apart from Exposed, would like to emerge. Just how JDBC is doing. I really miss the lack of JDBC outside the Java ecosystem.

**Comment by smyrgeorge** [comment]
@teras I agree with your last comment. We need to distinguish the low level APIs. In my implementation I provide low level DB access plus Listen/notify for postgres. If you want take a look

**Comment by BierDav** [comment]
With the upcoming R2DBC support we come a lot closer to get Exposed working with Native, because all of the nasty `java.sql` dependencies will get abstracted away from core 

I have prepared a overview of things that still need to be adapted and :

**Dependencies**
1. InputStreams and Buffers: https://github.com/Kotlin/kotlinx-io
2. ConcourentHashMap: https://github.com/CharlieTap/cachemap
3. BigDecimal: Move to `jvm` specific code. Unfortunately it is also used for WindowFunctions like PercentRank and CumeDist which need to be move to `jvm` only too for the first part.
4. TimeUnit, DateTimeFormatter: https://github.com/Kotlin/kotlinx-datetime
5. AtomicReference: Use the kotlin implementation
6. UUID: Needs to be moved to the jvm specifc code of `exposed-core` because in future there probabliy will be `kotlin.uuid` anyhow which will fix this issue on the long run: https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/uuid.md

**Technical Problems to overcome**
1. There is some use of advance reflection for Enums, removing the use of reflection might result in breaking changes
2. The newly introduced `ResultApi` needs special some special treatment so that it directly outputs `ExposedBlob`, `kotlin.Array`
3. `DatabaseDialectMetadata` uses `BigDecimal` for `version` which I didn't really understand why this is necessary, but we definitely need to look into a solution for this. This can't be that hard I guess.

**New Driver**
Of course `exposed-core` for native is only the first part. The second part would be also a driver that connects exposed to a database. A good starting point would be `androidx.sqlite` (https://developer.android.com/kotlin/multiplatform/sqlite) because it is already designed to work on all platforms, for other drivers there might be platform specific code necessary.


**Conculsion**
So the only thing that currently is in my opinion really a bit of a deal breaker is the required third party dependency for `ConcourentHashMap`. But maybe there is a more elegant solution than this that doesn't compromise on performance.

**Comment by smyrgeorge** [comment]
Great news really!

Also, for the underlying driver, especially for MySQL and PostreSQL, you could also check this repository: https://github.com/smyrgeorge/sqlx4k

It's db driver, that I've implemented and using in native targets. In the benches that I did, I saw very very good results 🙂

I could also provide a community based integration with exposed in the future, as soon as you provide an stable interface of course 🙂

**Comment by nathanfallet** [comment]
Is there anyone working on this? Or do I add it to the list of PRs I want to make? (made one yersterday on Koog to support Native targets for example, and a few days ago I started working on one on a RabbitMQ Ktor plugin too; so porting to Multiplatform is my thing now)

**Comment by BierDav** [comment]
Don't think so. I gave up 🥹

**Comment by nikolamin** [comment]
@nathanfallet I also need Exposed to be kmp'ified. If you start working on it, please share your fork i will do my best to contribute to it!

**Comment by JGuck** [comment]
Hey @nikolamin I provided a first pull request (#2629) which might provide a migration path for the core. It is not much but maybe a start....

**Comment by ApoloApps** [comment]
If this is not done, Ktor Server using K/N is not feasible right? If so, this is a good usecase imo

**Comment by BierDav** [comment]
You're right! Tough I must point out, that this is probably currently a very limited usecase, because if you want to go further than just a simply hobby backend you will probably need to stick to jvm anyways, because even ktor has a lot of jvm-only implementations and this is not even considering all of the other dependecies you might need.

But as the Kotlin Ecosystem expands this seams to be a attractive alternative to GraalVM, because K/N is a lot faster in compilation and setup, even tough it might currently not be as fast in execetion in some cases. K/N support would be a major step in this direction! So I'm really looking forward seeing the PR of @JGuck get merged.

**Comment by szymonpoltorak** [comment]
Are there any updates on this topic?

**Comment by teras** [comment]
> Are there any updates on this topic?

Since this topic came up again, I used my Easter vacations (and some days off) and decided to finalize a Kotlin multi-platform ORM I have been working on for a long time. It is pure Kotlin, with a thin C layer only for native database drivers. I am preparing for a public announcement, but before that, I'd be grateful if you could have a look at it and give me some feedback. From my point of view, it solves all my problems - I hope this might help you too. The link to the project is here: https://github.com/teras/stormify

What do you think?
